DDD-6.md

```diff
@@ -0,0 +1,100 @@
+
+## Operator Design & Requirements
+
+## Scope & Focus
+Do we want to create an operator for Debezium server specifically or more general Debezium operator?
+The opeator could potentially support
+
+- Debezium connector for DS
+- Debezium connector for KafkaConnect (can we somehow reuse UI backend?)
+- Debezium connector for Strimzi? Does it make sense to translate our resource to Strimzi?
+- Debezium connectors for DS via strimzi resources
+- Debezium UI deployment
+
+
+Initially will start with support for Debezium Server. However the resources should be designed with the broader scope in mind from beginning 
+
+
+### Operator SDK options
+
+The two primary options to consider her are Golang's Operator SDK (Golang) or Java Operator Framework (as Quarkus extension) . In case of a java-centered team such as our it makes sense to go with JODSK --  especially since JOSDK has quite active development. 
+
+
+### Custom Resource Proposal
+The following is a proposed example of `DebeziumServer` custom resource instance (in this). 
+
+```yaml
+apiVersion: debezium.io/v1alpha1
+kind: DebeziumServer
+metadata: 
+    name: my-debezium
+spec:
+    version: "2.2"
+    storage:
+      type: persistent 
+      claimName: ds-data-pvc
+    quarkus:
+      log.console.json: false
+    format:
+      value:
+        type: json
+        schemas.enable: false
+    transforms:
+      - name: test
+        type: com.example.TestTransform
+        predicate: test
+        negate: false
+        prop: 42
+    sink:
+      type: kafka
+      producer.bootstrap.servers: dbz-kafka-kafka-bootstrap:9092
+      producer.key.serializer: org.apache.kafka.common.serialization.StringSerializer
+      producer.value.serializer: org.apache.kafka.common.serialization.StringSerializer
+    source:
+      class: io.debezium.connector.mongodb.MongoDbConnector
+      mongodb.connection.string: mongodb://debezium:dbz@mongo.debezium.svc.cluster.local:27017/?replicaSet=rs0
+      database.history: io.debezium.relational.history.FileDatabaseHistory
+      tasks.max: 1
+      topic.prefix: demo
+      database.include.list: inventory
+      collection.include.list: inventory.customers
+      offset.storage.file.filename: /debezium/data/offsets.dat
+      offset.flush.interval.ms: 0 
+Status:
+    conditions:
+        Status: True
+        Type: Ready
+        Message: Server my-debezium is operational 
+```
+
+
+## Open Questions
+
+1) Do we need a service? What for?
+2) How to handle "Templating" to allow customization of deplyments outside of the scope allowed by the operator 
+    - This is needed due to reconciliation (JOSDK doesn't allow merge reconciliations)
+    -  It also solves a problem of different requirements between sinks (e.g. only some require environment variables)
+3) Dynamic vs Static cnfiguration typing
```

Comment:
Let's start with the dynamic approach. It is the easiest to implement.
To de-couple sink/operator development we might allow combination of both. Does it make sense to introduce a separate namespace already that will represent the dynamic configuration and will be the only option right now while later on we could introduce statics too?

Reply:
I'm not really sure what you mean by namespace in this context. The idea here is that the properties under sink are mostly just gathered into a map. 

In theory in future we could do something such as 

```yaml
sink:
  kafka:
    bootstraps: kafka:9092
```

so esentially we would introduce a typed property for each sink

Reply:
Or... thinking about it.. which is probably the best from implementation point of view. 

The initial version is going to be 

```yaml
sink:
  type: kafka
  config:
     prop1:
     prop2:
```

Once again, `config` is essentially an arbitrary dynamic sink type (and also technical equivalent of the `additional` filed)

And then later once typed configs are introduced

```yaml
sink:
  type: kafka # must be kafka for kafkaConfig
  kafkaConfig:
    bootstraps: kafka:9092
    additional: 
      prop1:
      prop2:
```
This gives as the best backwards compatibility and also allows us to use the `type` field as a decision point for which typed config to  read (if none of known values matched, then the dynamic `config` would be expected)

WDYT?